### PR TITLE
Researcher jwts

### PR DESCRIPTION
**Pull request by scytacki**
Allow project staff to request JWTs for specific target users.
This is done by passing both a resource_link_id and a target_user_id

If the current user is a portal admin they get the JWT.
If the current user is a project admin or researcher, they only get the JWT if the target_user has a permission form for a project that the current user is a admin or researcher of.

The claims of this kind of JWT looks like:
```
  "user_type": "user",
  "user_id": url_for_user,
  "platform_id": site_url,  // portal base URL ie https://learn.concord.org
  "platform_user_id": user.id,  // integer id of the current user 
  "class_hash": offering.class.class_hash,  // string from the resource_link_id's class
  "offering_id": offering.id, // integer of the offering should match the resource_link_id
  "target_user_id": target_user_id, // integer from the passed in parameter
```

If a resource_link_id is passed without a target_user_id and the current user is not a student or teacher of this resource_link then an error is returned.

If no resource_link_id is passed then the JWT for the user has the same claims as before:
```
  "user_type": "user",
  "user_id": url_for_user,
  "platform_id": site_url,  // portal base URL ie https://learn.concord.org
  "platform_user_id": user.id,  // integer id of the current user 
```

The intent of this new support is to allow researchers to view the work of students stored in external systems, specifically in the report-service.  We are now adding links to the researcher report (CSV file) that a researcher can click on and view the interactive with the student state loaded. These links are handled by the portal-report codebase. With this new JWT support the portal-report can request a JWT for the researcher which the report-service can use to grant the researcher access.

To complete the full circle of this work:
- the portal-report needs to be updated to send the target_user_id parameter when requesting the JWT
- the report-service firestore rules need to be updated to support these new JWTs. 

Firestore Rule support:
The most simple change is to treat these JWTs like teacher JWTs. The permission of a teacher JWT is granted based on the class_hash: 
https://github.com/concord-consortium/report-service/blob/e1394e676ffa835d413001cca31cc38fbb8f6117/firestore.rules#L22-L26
So that code could check for a user_type of "teacher" or "user"
The downside of the simple approach is that we are basically granting the researcher access to all students in the class.

The better approach is to use the target_user_id. And only grant the researcher/admin access to that specific learner. So that means a new call function here:
https://github.com/concord-consortium/report-service/blob/e1394e676ffa835d413001cca31cc38fbb8f6117/firestore.rules#L81-L85
Like `researcherOfOwner(resource)`, and a copy of this function:
https://github.com/concord-consortium/report-service/blob/e1394e676ffa835d413001cca31cc38fbb8f6117/firestore.rules#L30-L34

### CODE REVIEW

File: rails/app/controllers/api/v1/jwt_controller.rb [hunk 1]
```diff
@@ -145,39 +218,45 @@ def firebase
...
+
+        offering = Portal::Offering.find(resource_link_id)
+        class_hash = offering.clazz.class_hash
+        sub_claims.merge!({
+          class_hash: class_hash,
+          offering_id: offering.id,
+          # Any systems granting access based on this target_user_id claim should scope it to the
+          # class_hash and or offering_id so the main user doesn't gain access to all of the target user's
+          # data. Our policies don't always restrict access like this, but they should when possible.
+          target_user_id: params[:target_user_id].to_i
```
Comment (pjanik) [inline rails/app/controllers/api/v1/jwt_controller.rb:256]: I think we might end up with `target_user_id: 0` when params[:target_user_id] is `nil`. 
If there's a student with ID = 0, seems like a way to steal his data. 😉 
But seriously, maybe it's better to add this param to JWT only if it's explicitly provided in the request? I can imagine a developer inspecting JWT by hand in one of the client apps and wondering why there's some `target_user_id` defined at all. Well, I don't know if we can get to this point without `target_user_id` defined (auth failing somewhere earlier?), but just double checking.

**Review by pjanik — COMMENTED** [review]
Generally looks good to me. I've left a few inline comments/questions. 

The only thing that can really needs some update can be `return true if user.has_role?('manager','admin','researcher')`, but I might not understand that correctly. Other things seem minor.

**Review by pjanik — APPROVED** [review]

File: rails/app/controllers/api/v1/jwt_controller.rb [hunk 2]
```diff
@@ -39,33 +86,52 @@ def handle_initial_auth
...
+        end
+
+        if !can_access_user(user, params[:target_user_id])
+          raise StandardError, "Current user does not have permission to view target user."
+        end
+
+        # If the auth token included a teacher or student in it, we override them since
+        # we are going to use generic user access at this point
+        teacher = nil
+        learner = nil
```
[the remark this answers is not in the document]
Comment (scytacki) [reply rails/app/controllers/api/v1/jwt_controller.rb:130]: Yeah the method is overly complicated for backward compatibility. In some cases the access grant associated with the portal auth token has a learner and/or teacher set on it. 

In this case check_for_auth_token will return a role with the teacher and/or teacher set. Which will then be assigned to the teacher and learner variables on line 85 and 86.

So we need to ignore those variables when they are being overridden based on the url params. 

Ideally we'd get rid of these fields in the access_grant, they make this logic complex. But still most apps request a JWT without the resource_link_id.

File: rails/app/controllers/api/v1/jwt_controller.rb (hunk 1 above)
Comment (scytacki) [reply rails/app/controllers/api/v1/jwt_controller.rb:256]: Yeah it is confusing. I couldn't find a better way to handle all the cases. If the resource_link_id is set and the current user is not a student or teacher of that resrouce_link, then handle_initial_auth will raise an exception if target_user_id is not set. 

I can make this a bit more clear by adding that to the conditional on line 245 and adding a comment about the handle_initial_auth

**Pull request merged**